File: Guija/_OOC/AED/codigo_aed/src/ESTRUCTURAS_LINEARES_linear_estructures/4.COLAS (queues)/CircularQueue.c
/**
 * Implementacion de cola circular con arreglo.
 * Mantiene head, tail y tamano actual para O(1) en encolar/desencolar.
 * @author AxxturelHyphen
 */

#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>

typedef struct {
    int *arr;
    int capacity;
    int head;
    int tail;
    int current_size;
} CircularQueue;

static CircularQueue *create_queue(int capacity) {
    CircularQueue *q = (CircularQueue *)malloc(sizeof(CircularQueue));
    if (!q) {
        fprintf(stderr, "Sin memoria para la cola\n");
        exit(EXIT_FAILURE);
    }
    q->arr = (int *)malloc(sizeof(int) * capacity);
    if (!q->arr) {
        fprintf(stderr, "Sin memoria para el arreglo de la cola\n");
        free(q);
        exit(EXIT_FAILURE);
    }
    q->capacity = capacity;
    q->head = 0;
    q->tail = 0;
    q->current_size = 0;
    return q;
}

static bool is_empty(const CircularQueue *q) {
    return q->current_size == 0;
}

static bool is_full(const CircularQueue *q) {
    return q->current_size == q->capacity;
}

static bool enqueue(CircularQueue *q, int value) {
    if (is_full(q)) {
        return false; // overflow
    }
    q->arr[q->tail] = value;
    q->tail = (q->tail + 1) % q->capacity;
    q->current_size++;
    return true;
}

static bool dequeue(CircularQueue *q, int *out_value) {
    if (is_empty(q)) {
        return false; // underflow
    }
    *out_value = q->arr[q->head];
    q->head = (q->head + 1) % q->capacity;
    q->current_size--;
    return true;
}

static bool peek(const CircularQueue *q, int *out_value) {
    if (is_empty(q)) {
        return false;
    }
    *out_value = q->arr[q->head];
    return true;
}

static void free_queue(CircularQueue *q) {
    if (q) {
        free(q->arr);
        free(q);
    }
}

int main(void) {
    CircularQueue *q = create_queue(3);

    enqueue(q, 10);
    enqueue(q, 20);

    int value = 0;
    if (dequeue(q, &value)) {
        printf("%d\n", value); // 10
    }

    enqueue(q, 30);
    enqueue(q, 40); // ahora esta llena

    if (dequeue(q, &value)) printf("%d\n", value); // 20
    if (dequeue(q, &value)) printf("%d\n", value); // 30
    printf("%s\n", is_empty(q) ? "true" : "false"); // false
    if (dequeue(q, &value)) printf("%d\n", value); // 40
    printf("%s\n", is_empty(q) ? "true" : "false"); // true

    // peek de ejemplo
    enqueue(q, 50);
    if (peek(q, &value)) {
        printf("Front: %d\n", value);
    }

    free_queue(q);
    return 0;
}

File: Guija/_OOC/AED/codigo_aed/ESTRUCTURAS_LINEARES_linear_estructures/4.COLAS (queues)/CircularQueue.java
/**
 * Cola circular basada en array.
 * Reutiliza el arreglo de forma circular con indices head/tail mod capacity,
 * evitando movimientos de datos y mostrando el esquema clasico de buffer circular.
 * @author AxxturelHyphen
 */
class CircularQueue {
    private int[] arr;
    private int capacity;
    private int head;
    private int tail;
    private int currentSize;

    /**
     * Crea una cola circular de capacidad fija.
     * @param N cantidad maxima de elementos encolados
     */
    CircularQueue(int N) {
        capacity = N;
        arr = new int[N];
        head = 0;
        tail = 0;
        currentSize = 0;
    }

    /**
     * Encola un valor al final de la cola.
     * @param value dato a guardar
     * @throws Exception si la cola esta llena
     */
    void enqueue(int value) throws Exception {
        if (currentSize == capacity) {
            throw new Exception("Queue overflow (full)");
        }
        arr[tail] = value;
        tail = (tail + 1) % capacity; // avanza circularmente
        currentSize++;
    }

    /**
     * Desencola y devuelve el elemento frontal.
     * @return valor almacenado en la posicion head
     * @throws Exception si la cola esta vacia
     */
    int dequeue() throws Exception {
        if (currentSize == 0) {
            throw new Exception("Queue underflow (empty)");
        }
        int value = arr[head];
        head = (head + 1) % capacity; // avanza circularmente
        currentSize--;
        return value;
    }

    /**
     * Devuelve el elemento frontal sin retirarlo.
     * @return valor en head
     * @throws Exception si la cola esta vacia
     */
    int peek() throws Exception {
        if (currentSize == 0) {
            throw new Exception("Empty queue");
        }
        return arr[head];
    }

    /** Indica si la cola no tiene elementos. */
    boolean isEmpty() {
        return currentSize == 0;
    }

    /** Indica si la cola alcanzo la capacidad declarada. */
    boolean isFull() {
        return currentSize == capacity;
    }

    static void main(){
        CircularQueue q = new CircularQueue(3);

        q.enqueue(10);
        q.enqueue(20);
        System.out.println(q.dequeue()); // 10

        q.enqueue(30);
        q.enqueue(40);

        System.out.println(q.dequeue()); // 20
        System.out.println(q.dequeue()); // 30
        System.out.println(q.isEmpty()); // false
        System.out.println(q.dequeue()); // 40
        System.out.println(q.isEmpty()); // true
    }
}
